### ISSUE TITLE
Use Prisma Client from within the browser to connect to SqlLite Javascript in memory database without server side

### ISSUE THREAD

**Issue by PhilAndrew** [prisma/prisma#5385]
## Problem

I wish to use Prisma Client from within the browser entirely without a server, connecting to SqlLite Javascript in memory, when I tried to pack it up with browserify it complained in the console log that it would not work in the browser.

> Example
> const client = new PrismaClient()
> Throws
> Error: PrismaClient is unable to be run in the browser. In case this error is unexpected for you, please report it in https://github.com/prisma/prisma-client-js/issues

The motivation for doing this is that the trend is to move towards JAMStack pages which are webpages where most of the functionality is HTML and Javascript and minimizing the server side. If I can load a SqlLite database in Javascript in the client browser from static files hosted on CDN then it makes it easy to provide the type of data querying I want to do at a lower cost (no server required) for a large number of users. That is, I can scale out to many users without the need for a Prisma server side. I could query a SqlLite database which is just a file. I could send any changes in GraphQL off to be processed to somewhere else.

SqlLite in Javascript is here
https://github.com/sql-js/sql.js

JamStack sites are now getting popular.
https://jamstack.org/

## Suggested solution

Well I'm not sure if it can even be done. I see TypeORM can work with SqlLite in Javascript all within the browser and I think that's useful.

**Comment by nirajhinge** [comment]
We have a similar use where we want to use Prisma with sqlite from within an app built with Ionic/Vue. Individual apps will have their own the sqlite database and when we push for updates, if there are database changes, we will run the migration on each of those devices. If it's the initial installation, it will simply run all the migrations. This would add a great usage for things like offline apps which store the data in databases like sqlite and then sync when online.

**Comment by janpio** [comment]
Quick background why this would be difficult in the current implementation of our Query Engine:

The Query Engine we are using ships with the SQLite client and can only talk to a local file. That is a limitation that makes total sense on the server side where Prisma is usually used. It becomes harder in a web/browser context, where there is not real file system and SQLite is usually provided via a browser API or a native API (as in the Capacitor and Cordova plugin). I am sure there is a solution on how to make them talk to each other, but it is not something that works out of the box and that will be trivial to figure out and fix (Especially with all the mobile platforms, versions, changes by device manufacturer etc)

**Comment by bencompton** [comment]
My use case is to use Node.js + MySQL in prod, but SQL.js in the browser during development with the Node.js back-end APIs running in the browser. Integration tests would be running the front-end and back-end together with SQL.js as well. So, development could mostly occur in the browser with this type of setup.

As it stands now, the query builder and query engine are coupled together in the Prisma client. However, I wonder if Prisma could have a client type that doesn't even use an engine binary, and instead calls a user-defined callback. For example, imagine if queries could be routed to SQL.js (or other JavaScript libraries) like so:

```javascript
const db = new SQL.Database();

const prisma = new EnginelessPrismaClient((sql: string) => {
  return db.exec(sql);
});

const allUsers = await prisma.user.findMany({
  include: { posts: true }
});
```

Perhaps the callback would have to be way more complex than I am imagining, or there is some reason why this wouldn't even be possible.

**Comment by biels** [comment]
I'm also interested in this use case. For node.js there are lots of alternatives already. For use in a JS environment w/ SQL.js / sqlite there is only TypeORM currently. 

Prisma has the option to establish itself as a viable alternative for that use case easily. 
Especially considering the trend towards edge computing and offline first and the fact that the expected user experience is making a database on frontends more common by the day.

Using the same rust engine compiled to WASM seems like a good idea to me.

**Comment by crubier** [comment]
Up on this topic! There are now examples of SQLite running in the browser in WASM on top of Origin-Private FileSystem (OPFS). 

This should make a browser based version of Prisma doable, and extremely appealing, for offline use cases.

https://news.ycombinator.com/item?id=33374402

**Comment by crubier** [comment]
Chrome dev blog mentioning this approach today, this is gaining steam! Would be awesome to get Prisma support

https://developer.chrome.com/blog/sqlite-wasm-in-the-browser-backed-by-the-origin-private-file-system/

**Comment by danielrotaermel** [comment]
Seems like the prisma vscode extension already uses a WASM engine.
https://www.prisma.io/blog/vscode-extension-prisma-rust-webassembly
https://www.prisma.io/blog/wnip-q4-dsk0golh8v#vs-code-extension-now-uses-a-wasm-module

**Comment by biels** [comment]
> Seems like the prisma vscode extension already uses a WASM engine. https://www.prisma.io/blog/vscode-extension-prisma-rust-webassembly https://www.prisma.io/blog/wnip-q4-dsk0golh8v#vs-code-extension-now-uses-a-wasm-module

Exactly, since the engine is written in Rust it is quite easy to build a WebAssembly bundle to enable support for the browser. This is a huge untapped potential for prisma. Now that OPFS is available, the recommended way of doing SQL in the browser is already via wasm. Hopefully someone from the team can take a look!

**Comment by danielrotaermel** [comment]
Yeah, the potential is huge. For example you could build offline first PWAs and use the same prisma client on a server and directly browser.

**Comment by mathieu-heyarctic** [comment]
the prisma-schema engine code is released here: 
https://github.com/prisma/prisma-engines/blob/main/prisma-schema-wasm

They also have a wasm engine here:
https://github.com/prisma/prisma-engines/tree/main/query-engine/query-engine-wasm

so it seems like the plumbing is available- and the main repo has gotten lots of WASM attention in the past few weeks.

**Comment by matbee-eth** [comment]
the prisma-schema engine code is released here:
https://github.com/prisma/prisma-engines/blob/main/prisma-schema-wasm

They also have a wasm query engine here:
https://github.com/prisma/prisma-engines/tree/main/query-engine/query-engine-wasm

so it seems like the plumbing is available- and the main repo has gotten lots of WASM attention in the past few weeks.

Our best bet is to compile each wasm module and run the tests as a web client instead of a server node client

**Comment by archywillhe** [comment]
any updates on this? also related - how's wasm-support for prisma client now ? I saw there were some binaries needed so wasm is not fully et

**Comment by juemrami** [comment]
> Yeah, the potential is huge. For example you could build offline first PWAs and use the same prisma client on a server and directly in the browser.

This is exactly my use case.

It would make migrating any pre-existing _non_ local-first projects a breeze. Youd be comfortable knowing you'd have the same type contracts when building out your local sqlite database to match your server.

**Comment by bselwayregis** [comment]
+1 this... would be amazing

**Comment by isimmons** [comment]
Would be great for building Tauri apps. There is a rust plugin for doing things on the rust backend but being able to keep an sqlite database on the front end would greatly simplify the creation of desktop apps with more than just localstorage for offline.

**Comment by gregpalaci** [comment]
+1 https://github.com/prisma/prisma/issues/5385#issuecomment-2417881298
Would be hella cool to use with [service-workers / hono](https://hono.dev/docs/getting-started/service-worker)
```
app.get(
  "/users",
  cache({
    cacheName: "users",
    wait: true,
    cacheControl: "max-age=3600",
  }),
  async (c) => {
    const referer = c.req.header("Referer");
    c.header("Access-Control-Allow-Origin", "*");
    c.header("Access-Control-Allow-Methods", "OPTIONS,GET,PUT,POST,DELETE");
    c.header("Access-Control-Allow-Headers", "*");

    console.log(c.req.routePath.replace("sw/", ""));
    const data = await ofetch(
      "http://localhost:3010" + c.req.routePath.replace("sw/", "")
    );
    console.log({ data });
    return c.json(data);
  }
);
```

currently CORS isnt an issue, but have to forward a SW.js request to a server 🤦

**Comment by hichemfantar** [comment]
Running prisma in the browser would be amazing. i could def use this feature for local first apps

**Comment by jswhisperer** [comment]
https://dexie.org/#sync

dexie is neat too

**Comment by hichemfantar** [comment]
Dexie uses indexeddb.

I find sql.js (sqlite wasm) and drizzle to be a nice combo in the browser

**Comment by clicktodev** [comment]
considering prisma 7 client is rust free, i believe it should be possible to use prisma in the browser now

**Comment by banou26** [comment]
It has been possible for months.
I successfully used Prisma in the browser around 6months ago.
Requires a lot of hacking though.
Also the performance is really not great, the query builder is incredibly slow, so I ended up switching to SQLite and drizzle instead.

You can browse around to see how it integrates.
https://github.com/Banou26/stub/tree/05bdef59126aff71ec21b88bfd81dc5fbb153b99/src/worker/prisma

**Comment by rich-row-chat** [comment]
Prisma has a WASM query builder and now makes this practical, so we published: @row-chat/prisma-sqlite-wasm-adapter, an in-browser Prisma adapter for @sqlite.org/sqlite-wasm.

- Source: https://github.com/row-chat/prisma-browser-adapters                                                                                   
- Live demo: https://row-chat.github.io/prisma-browser-adapters/ — the Chinook sample database running entirely in the browser, exposed through a Prisma Client REPL, a raw SQL REPL, and an embedded Prisma Studio. SQLite-WASM persists the database in OPFS via the SAH pool VFS.
                                 
How it works: the sqlite-wasm DB lives in a SharedWorker (so all tabs on the same origin see the same database). Queries like prisma.album.findMany() work entirely in the browser.

**Comment by rich-row-chat** [comment]
We also have working in-browser adapters for wa-sqlite and PGLite that we plan to publish in the future.

**Comment by rich-row-chat** [comment]
The Prisma adapter for wa-sqlite has been published: https://www.npmjs.com/package/@row-chat/prisma-wa-sqlite-adapter

**Comment by rich-row-chat** [comment]
PGLite can be used with Prisma in the browser without extra dependencies. Here are the instructions. With the plain PGLite Prisma adapter each browser tab uses a full instance of PGLite. In the future the package described in this link will be published to enable one instance of PGLite to coordinate through a SharedWorker: https://github.com/row-chat/prisma-browser-adapters/tree/main/packages/prisma-pglite-adapter#readme
